src/ray/id.h

```diff
@@ -2,44 +2,128 @@
 #define RAY_ID_H_
 
 #include <inttypes.h>
+#include <limits.h>
 
+#include <chrono>
 #include <cstring>
+#include <mutex>
+#include <random>
 #include <string>
 
 #include "plasma/common.h"
 #include "ray/constants.h"
+#include "ray/util/logging.h"
 #include "ray/util/visibility.h"
 
 namespace ray {
 
-class RAY_EXPORT UniqueID {
+class DriverID;
+class UniqueID;
+
+// Declaration.
+std::mt19937 RandomlySeededMersenneTwister();
+uint64_t MurmurHash64A(const void *key, int len, unsigned int seed);
+
+// Change the compiler alignment to 1 byte (default is 8).
+#pragma pack(push, 1)
+
+template <typename T>
+class BaseID {
  public:
-  UniqueID();
-  UniqueID(const plasma::UniqueID &from);
-  static UniqueID from_random();
-  static UniqueID from_binary(const std::string &binary);
-  static const UniqueID &nil();
+  BaseID();
+  static T from_random();
+  static T from_binary(const std::string &binary);
+  static const T &nil();
+  static size_t size() { return T::size(); }
+
   size_t hash() const;
   bool is_nil() const;
-  bool operator==(const UniqueID &rhs) const;
-  bool operator!=(const UniqueID &rhs) const;
+  bool operator==(const BaseID &rhs) const;
+  bool operator!=(const BaseID &rhs) const;
   const uint8_t *data() const;
-  static size_t size();
   std::string binary() const;
   std::string hex() const;
-  plasma::UniqueID to_plasma_id() const;
 
- private:
+ protected:
+  BaseID(const std::string &binary) {
+    std::memcpy(const_cast<uint8_t *>(this->data()), binary.data(), T::size());
+  }
+  // All IDs are immutable for hash evaluations. mutable_data is only allow to use
+  // in construction time, so this function is protected.
+  uint8_t *mutable_data();
+  // For lazy evaluation, be careful to have one Id contained in another.
+  // This hash code will be duplicated.
+  mutable size_t hash_ = 0;
+};
+
+class UniqueID : public BaseID<UniqueID> {
+ public:
+  UniqueID() : BaseID(){};
+  static size_t size() { return kUniqueIDSize; }
+
+ protected:
   UniqueID(const std::string &binary);
 
  protected:
   uint8_t id_[kUniqueIDSize];
-  mutable size_t hash_ = 0;
 };
 
-static_assert(std::is_standard_layout<UniqueID>::value, "UniqueID must be standard");
+class TaskID : public BaseID<TaskID> {
+ public:
+  TaskID() : BaseID() {}
+  static size_t size() { return kTaskIDSize; }
+  static TaskID GetDriverTaskID(const DriverID &driver_id);
+
+ private:
+  uint8_t id_[kTaskIDSize];
+};
+
+class ObjectID : public BaseID<ObjectID> {
```

Comment:
do we still require ObjectID to be 20 bytes with this change? e.g. if we add some bits into ObjectID, is there any place other than the ObjectID structure itself that would need to be changed?

Reply:
Yes, ObjectID still requires 20 bytes, but you can specify some bits later on. For example, you can use the following code:
```c
private:
  uint8_t id_[kTaskIDSize];
  int32_t is_put : 1;
  int32_t other_flag : 1;
  int32_t index: 30;
```

Reply:
@guoyuhong got it, thanks. I'm feeling the flags may potentially require more bits thus flags + index might not fit into an uint32_t. Would it be possible to make it an uint64_t instead, would 12 bytes be sufficient for TaskID?

Reply:
In current ArcOS design, there are 2 bytes used for AppId, 4 extra bytes for BatchId, 8 extra bytes for TaskId, 6 extra bytes for ObjectID. The schema is not determined. I will propose a schema, if they do not give out a schema by the end of this week.

Reply:
@guoyuhong Thanks for the explanations. Yes I think 6 bytes for flags + index should be sufficient, was just not convinced if 4 bytes would work:)